lib/mayaUsd/undo/UsdUndoManager.cpp

```diff
@@ -29,9 +29,16 @@ UsdUndoManager& UsdUndoManager::instance()
     return undoManager;
 }
 
-void UsdUndoManager::trackLayerStates(const SdfLayerHandle& layer)
+void UsdUndoManager::trackStatesOnNewStage(const SdfLayerHandle& layer)
 {
-    layer->SetStateDelegate(UsdUndoStateDelegate::New());
+    _layerStateDelegate = TfCreateRefPtr(new UsdUndoStateDelegate);
+
+    layer->SetStateDelegate(_layerStateDelegate);
+}
+
+void UsdUndoManager::trackStatesOnEditTargetChange(const SdfLayerHandle& layer) 
```

Comment:
I'm a bit perplexed here.  If we set a new state delegate on layer creation, then why do we have to set it on edit target change?  Is it because of initialization order and the fact that certain layers may have been created before the UsdUndoManager starts tracking changes?

Reply:
Every time we change the target edit layer, UsdUndoStateDelegate::_OnSetLayer is invoked where it
needs to set the _layer for tracking changes. 

We do need to call layer->SetStateDelegate() when we switch between layers otherwise how would the layer know what state delegate needs to use?

Notice that in trackStatesOnEditTargetChange, I am only passing the _layerStateDelegate object that was created on stage set. 

More than one layer needs to share the same StateDelegate .
```
void UsdUndoManager::trackStatesOnEditTargetChange(const SdfLayerHandle& layer) 
{
    layer->SetStateDelegate(_layerStateDelegate);
}
```